docs/source/app-dev/command-deduplication.rst

```diff
@@ -0,0 +1,339 @@
+
+
+Known processing time bounds
+============================
+
+For now, we assume that the application knows a bound ``B`` on the processing time and forward clock drifts (against the application’s clock) in the Daml ledger.
+If processing measured across all retries takes longer than the bound ``B``, the ledger change may take effect several times.
+Under this assumption, the following strategy works for applications that use the :ref:`Command Service <command-service>` or the :ref:`Command Submission <command-submission-service>` and :ref:`Command Completion Service <command-completion-service>`.
+
+.. _dedup-bounded-step-command-id:
+
+#. Choose a fresh command ID for the ledger change and the ``actAs`` parties, which (together with the application ID) determine the change ID.
+   Remember the command ID across application crashes.
+
+   .. note::
+      Make sure that you set command ID deterministically, that is to say: the same ledger change must use the same command ID for all command submissions.
+      This is useful for the recovery procedure after an application crash/restart, in which the application inspects the state of the ledger (e.g., via the :ref:`Active contracts service <active-contract-service>`) and sends commands to the ledger.
+      When using deterministic command IDs, any commands that had been sent before the application restart will be discarded by the ledger to avoid duplicate submissions.
+
+   .. _dedup-bounded-step-offset:
+
+#. When you use the :ref:`Command Completion Service <command-submission-service>`, obtain a recent offset on the completion stream ``OFF1``, say the :ref:`current ledger end <com.daml.ledger.api.v1.CommandCompletionService.CompletionEnd>`.
+
+   .. _dedup-bounded-step-submit:
+   
+#. Submit the command with the following parameters
+
+   - Set the :ref:`command ID <<com.daml.ledger.api.v1.Commands.command_id>>` to the chosen command ID from :ref:`step 1 <dedup-bounded-step-command-id>`.
+
+   - Set the :ref:`deduplication duration <com.daml.ledger.api.v1.Commands.deduplication_duration>` to the bound ``B``.
+
+   - Set the :ref:`submission ID <com.daml.ledger.api.v1.Commands.submission_id>` to a fresh value, e.g., a random UUID.
+
+   - Set the timeout (gRPC deadline) to the expected processing (Command Service) or submission (Command Submission Service) delay.
+
+     The **processing delay** measures the time between when the application sends off a submission to the :ref:`Command Service <command-service>` and when it receives the acceptance or rejection.
+     The **submission delay** measures the time between when the application sends off a submission to the :ref:`Command Submission Service <command-submission-service>` and when it obtains a synchronous response for this gRPC call.
+     After the RPC timeout, the application considers the submission as lost and enters a retry loop.
+     This timeout is typically much shorter than the deduplication duration.
```

Comment:
Yes. You get some gRPC response back from the asynchronous command submission service as well.

Reply:
I think this would clear my confusion:

```suggestion
   - Set the timeout (gRPC deadline) to the expected processing (Command Service) or submission (Command Submission Service) delay.

     The **processing delay** measures the submission processing time, i.e. the time between when the application sends off a submission to the :ref:`Command Service <command-service>` and when it receives (synchronously, unless it times out) the acceptance or rejection.
     The **submission delay** measures the submission hand-off time, i.e. the time between when the application sends off a submission to the :ref:`Command Submission Service <command-submission-service>` and when it obtains a synchronous response for this gRPC call.
     After the RPC timeout, the application considers the submission as lost and enters a retry loop.
     This timeout is typically much shorter than the deduplication duration.
```